### PR TITLE
Improvements for RTL

### PR DESCRIPTION
**Pull request by Shreeshrii**

### CODE REVIEW

File: generate_wordstr_box.py [hunk 1]
```diff
@@ -44,6 +38,9 @@ def trim(im):
 # create WordStr line boxes for Indic & RTL
 for line in lines:
     line = unicodedata.normalize('NFC', line.strip())
+    if args.rtl:
+        line = line.translate(str.maketrans("()[]{}»«><", ")(][}{«»<>"))
```
Comment (stweil) [inline generate_wordstr_box.py:42]: Are you sure that translation is needed?

File: generate_wordstr_box.py [hunk 2]
```diff
@@ -44,6 +38,9 @@ def trim(im):
...
+        line = bidi.algorithm.get_display(line)
```
Comment (stweil) [inline generate_wordstr_box.py:43]: I made a short test and think that `bidi.algorithm.get_display(line)` could be used unconditionally below. That would simplify the handling a lot.

File: generate_wordstr_box.py [hunk 3]
```diff
@@ -44,6 +38,9 @@ def trim(im):
...
     if line:
         print("WordStr 0 0 %d %d 0 #%s" % (width, height, line))
-        print("\t 0 0 %d %d 0" % (width, height))
+        print("\t 0 0 %d %d 0" % (width, height), end='')
```
Comment (stweil) [inline generate_wordstr_box.py:46]: This suppresses the LF, so the file size is reduced by one byte. Does it also have an effect on the training?

File: generate_wordstr_box.py [hunk 4]
```diff
@@ -16,26 +17,19 @@
...
+# Reading direction
+arg_parser.add_argument('-r', '--rtl', action='store_true', default=False, help='RTL input')
+
 args = arg_parser.parse_args()
 
 #
 # main
 #
 
-def trim(im):
```
Comment (stweil) [inline generate_wordstr_box.py:25]: So you no longer suggest to use trimming. Why?

File: generate_wordstr_box.py (hunk 1 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:42]: No, I am not sure. 

I had tried initially without translation - please see

https://github.com/Shreeshrii/tesstrain-ckb

https://github.com/Shreeshrii/tesstrain-ckb/blob/master/generate_wordstr_box.py

The accuracy reports shows low levels of recognition for these punctuation marks.

https://github.com/Shreeshrii/tesstrain-ckb/blob/master/reports/ckb-eval-ckbLayer_1.413_255615_562800/report_ckbLayer_1.413_255615_562800-ckb-ALL.txt#L462

Also while doing lstmtraining with start_model tessdata_best/script/Arabic, without translation the debug lines shows the punctuation marks in opposing direction for the OCR text and Ground truth, with translation they match.

Hence I thought that approach was incorrect.

File: generate_wordstr_box.py (hunk 2 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:43]: Yes, that would simplify.

File: generate_wordstr_box.py (hunk 3 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:46]: I was getting some unexpected results and hence had added that. It might have been when I was using files generated by text2image.

File: generate_wordstr_box.py (hunk 4 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:25]: No need to add additional overhead of trimming specially when dealing with thousands of images. Assumption is that images and ground-truth are pre-processed to meet training needs.

**Comment by stweil** [comment]
@Shreeshrii, I suggest to separate two of the changes in two new pull requests:

1. pull request for PSM 13
2. pull request which removes trimming

Those could be merged immediately.

File: generate_wordstr_box.py (hunk 4 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:25]: Done via https://github.com/tesseract-ocr/tesstrain/pull/160

File: generate_wordstr_box.py (hunk 3 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:46]: Now I remember, it was needed for the ground-truth text files. 
Not required for box files.

File: generate_wordstr_box.py (hunk 1 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:42]: With translation

Iteration 63: GROUND  TRUTH : ییاڵووقەل ییراکەوەنیزۆد و ناڕەگ یراک ۆب ی ) ةصاوغ ( ییایرەد رێژ ەک ەووب شیسەک نیمەکەی اهەورەه ،ەوەییزۆد
Iteration 63: BEST OCR TEXT : ىياڵووقەل ىیراكەوەنیزۆد و ناڕەگ یراک ۆب ى ) ةصاوغ ( ىيایرەد رێژ ەک ەووب شیسەك نیەەکەي اھەورەه .ەوەيیزۆد
File data/rtltest-ground-truth/000103.lstmf line 0 :

I will rerun the rtltest with both translation and no translation and share results.

Comment (stweil) [reply generate_wordstr_box.py:42]: Thanks.

File: generate_wordstr_box.py (hunk 3 above)
Comment (stweil) [reply generate_wordstr_box.py:46]: So ground truth text files must not end with a line feed? What happens if they do? Some editors add a line feed each time when a file was changed and saved.

File: generate_wordstr_box.py (hunk 1 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:42]: In my test, with the trainingdata in rtltest.tgz `with translation` and `without translation`, the option `with translation` gives better results. The training data only had `( and )` , not the other punctuation.

rtltest - with translation
```
    6038   Characters
     383   Errors
   93.66%  Accuracy

   Count   Missed   %Right
     837       99    88.17   ASCII Spacing Characters
     107        2    98.13   ASCII Special Symbols
      32        0   100.00   ASCII Digits
    3139      164    94.78   Basic Arabic
    1923       78    95.94   Arabic Extended
    6038      343    94.32   Total


   Count   Missed   %Right
     111        0   100.00   {<\n>}
     726       99    86.36   { }
       1        0   100.00   {"}
      23        1    95.65   {(}
      19        0   100.00   {)}
       8        0   100.00   {-}
      38        1    97.37   {.}
       4        0   100.00   {/}
```

notranslate - without translation

```
    6038   Characters
     421   Errors
   93.03%  Accuracy


   Count   Missed   %Right
     837       99    88.17   ASCII Spacing Characters
     107       43    59.81   ASCII Special Symbols
      32        0   100.00   ASCII Digits
    3139      166    94.71   Basic Arabic
    1923       77    96.00   Arabic Extended
    6038      385    93.62   Total


   Count   Missed   %Right
     111        0   100.00   {<\n>}
     726       99    86.36   { }
       1        0   100.00   {"}
      23       23     0.00   {(}
      19       19     0.00   {)}
       8        0   100.00   {-}
      38        1    97.37   {.}
       4        0   100.00   {/}
```

Comment (Shreeshrii) [reply generate_wordstr_box.py:42]: @stweil Please see if you can replicate this with my test data, as well as with the other training data that you have.

**Comment by Shreeshrii** [comment]
[accuracy-Arabic.traineddata-list.eval.txt](https://github.com/tesseract-ocr/tesstrain/files/4498357/accuracy-Arabic.traineddata-list.eval.txt)
[accuracy-notranslate.traineddata-list.eval.txt](https://github.com/tesseract-ocr/tesstrain/files/4498359/accuracy-notranslate.traineddata-list.eval.txt)
[accuracy-rtltest.traineddata-list.eval.txt](https://github.com/tesseract-ocr/tesstrain/files/4498360/accuracy-rtltest.traineddata-list.eval.txt)

I ws not able to add the accuracy reports in comment above.

File: generate_wordstr_box.py (hunk 1 above)
Comment (Shreeshrii) [reply generate_wordstr_box.py:42]: https://github.com/tesseract-ocr/tesstrain/pull/159#issuecomment-616024684
has links to the accuracy reports. I was not able to upload them in this thread.

File: Makefile [hunk 5]
```diff
@@ -70,16 +70,16 @@ else
 ifeq ($(LANG_TYPE),RTL)
 	NORM_MODE =3
 	RECODER =--pass_through_recoder --lang_is_rtl
-	GENERATE_BOX_SCRIPT =generate_wordstr_box.py
+	GENERATE_BOX_SCRIPT =generate_wordstr_box.py --rtl
 else
-	NORM_MODE =2
+	NORM_MODE =1
```
Comment (stweil) [inline Makefile:75]: This matches `src/training/language-specific.sh` and looks good, so it could also be moved to a separate pull request.

File: Makefile [hunk 6]
```diff
@@ -70,16 +70,16 @@ else
 ifeq ($(LANG_TYPE),RTL)
 	NORM_MODE =3
```
Comment (stweil) [inline Makefile:71]: @Shreeshrii, where does this norm mode come from? Should it be `NORM_MODE = 2`?

Comment (Shreeshrii) [reply Makefile:71]:
```
 unicharset_extractor

Usage: unicharset_extractor [--output_unicharset filename] [--norm_mode mode] box_or_text_file [...]
Where mode means:
 1=combine graphemes (use for Latin and other simple scripts)
 2=split graphemes (use for Indic/Khmer/Myanmar)
 3=pure unicode (use for Arabic/Hebrew/Thai/Tibetan)
Reads box or plain text files to extract the unicharset.

```

Comment (stweil) [reply Makefile:71]: Strange. That code was added by Ray in commit https://github.com/tesseract-ocr/tesseract/commit/a912967cc. The same commit also added the code which sets a different value in `src/training/language-specific.sh`. There was even a comment from you asking why the value was not set to 3.

Also strange: `unicharset_extractor --help` shows a different help text.

Comment (Shreeshrii) [reply Makefile:71]: One way to check will be to look at the lstm-unicharset files in `ara.traineddata` and other RTL languages.

Comment (stweil) [reply Makefile:71]: For my test data (Arabic handwriting, about 2500 lines) it did not make a difference whether `NORM_MODE` was 2 or 3. This can be explained by the Tesseract code which increments values below 3, so 2 becomes 3 during the execution.

**Comment by Shreeshrii** [comment]
@stweil So what is your recommendation regarding norm_mode? 
2 for Indic and RTL
1 for others
???

What about punctuation translation for RTL?
